base/compiler/typeinfer.jl

```diff
@@ -883,11 +895,24 @@ end
 
 #### entry points for inferring a MethodInstance given a type signature ####
 
+function maybe_synthesize_optimized_src!(interp::AbstractInterpreter, frame::InferenceState, run_optimizer::Bool)
+    if run_optimizer && may_discard_trees(interp) && is_result_constabi_eligible(frame.result)
+        # We didn't bother optimizing this code, because we determined early
+        # that we would use const abi. However, for consistency, we want our
+        # callers to feel assured that the code was eliminated, so rewrite the
+        # code now.
+        frame.src.code = Any[ReturnNode(quoted(frame.result.result.val))]
```

Comment:
We have a more correct version of this already in `typeinf_ext`. Could we move that into a helper function and use that here?